File: pdfservices/nodejs/compress_and_linearize.mjs
import fs from 'fs';
import { Readable } from 'stream';
import { finished } from 'stream/promises';
import path from 'node:path';

const CLIENT_ID = process.env.CLIENT_ID;
const CLIENT_SECRET = process.env.CLIENT_SECRET;
const HOST = process.env.HOST;

async function uploadDoc(filepath, id, secret) {
	let filename = path.basename(filepath);
	let data = new FormData();
	let blob = await fs.openAsBlob(filepath);
	data.append('file', blob, filename);

	let req = await fetch(`${HOST}/pdf-services/api/documents/upload`, {
		method:'POST',
		headers: {
			client_id: id, 
			client_secret: secret
		},
		body: data
	});

	return await req.json();

}

async function compressPDF(doc, id, secret, level="MEDIUM") {

	let headers = {
		"client_id":id,
		"client_secret":secret,
		"Content-Type":"application/json"
	};

	let body = {
		"documentId":doc,
		"compressionLevel":level 	
	};

	let resp = await fetch(`${HOST}/pdf-services/api/documents/modify/pdf-compress`, {
		method:"POST",
		headers, 
		body:JSON.stringify(body)
	});

	return await resp.json();

}

async function linearizePDF(doc, id, secret, level="MEDIUM") {

	let headers = {
		"client_id":id,
		"client_secret":secret,
		"Content-Type":"application/json"
	};

	let body = {
		"documentId":doc,
		"compressionLevel":level 	
	};

	let resp = await fetch(`${HOST}/pdf-services/api/documents/optimize/pdf-linearize`, {
		method:"POST",
		headers, 
		body:JSON.stringify(body)
	});

	return await resp.json();

}

async function checkTask(task, id, secret) {

	while(true) {
		let req = await fetch(`${HOST}/pdf-services/api/tasks/${task}`, {
			method:'GET',		
			headers: {
				client_id: id, 
				client_secret: secret,
				'Content-Type':'application/json'
			}
		});

		let result = await req.json();

		if(result['status'] === 'COMPLETED') {
			return result;
		} else if(result['status'] === 'FAILED') {
			console.log('Failure. Here is the last status:');
			console.log(result);
			process.exit(1);
		} else {
			console.log(`Current status, ${result['status']}, percentage: ${result['progress']}`);
			await delay(5);
		}

	}
}

// Lame function to add a delay to my polling calls
async function delay(x) {
	return new Promise(resolve => {
		setTimeout(() => resolve(), x * 1000);
	});
}

async function downloadResult(doc, filePath, id, secret) {
	let headers = {
		"client_id":id,
		"client_secret":secret
	}

	let res = await fetch(`${HOST}/pdf-services/api/documents/${doc}/download`, {
		headers
	});

	const body = Readable.fromWeb(res.body);
	const download_write_stream = fs.createWriteStream(filePath);
	return await finished(body.pipe(download_write_stream));

}

let input = "../../inputfiles/input.pdf";
console.log(`File size of input: ${fs.statSync(input).size}`)

let doc = await uploadDoc(input, CLIENT_ID, CLIENT_SECRET);
console.log(`Uploaded doc to Foxit, id is ${doc['documentId']}`);

let task = await compressPDF(doc.documentId, CLIENT_ID, CLIENT_SECRET, "HIGH");
console.log(`Created task, id is ${task['taskId']}`);

let result = await checkTask(task.taskId, CLIENT_ID, CLIENT_SECRET);
console.log("Done converting to PDF. Now doing linearize.")

task = await linearizePDF(result.resultDocumentId, CLIENT_ID, CLIENT_SECRET, "HIGH");
console.log(`Created task, id is ${task['taskId']}`);

result = await checkTask(task.taskId, CLIENT_ID, CLIENT_SECRET);
console.log("Done with linearize task.");


let output = "../../output/node_really_optimized.pdf";
downloadResult(result["resultDocumentId"], output , CLIENT_ID, CLIENT_SECRET);
console.log(`Done and saved to: ${output}.`)
console.log(`File size of output: ${fs.statSync(output).size}`)

File: pdfservices/python/compress_and_linearize.py
import os
import requests
import sys 
from time import sleep 

CLIENT_ID = os.environ.get('CLIENT_ID')
CLIENT_SECRET = os.environ.get('CLIENT_SECRET')
HOST = os.environ.get('HOST')

def uploadDoc(path, id, secret):
	
	headers = {
		"client_id":id,
		"client_secret":secret
	}

	with open(path, 'rb') as f:
		files = {'file': f}

		request = requests.post(f"{HOST}/pdf-services/api/documents/upload", files=files, headers=headers)
		return request.json()

def compressPDF(doc, level, id, secret):
	
	headers = {
		"client_id":id,
		"client_secret":secret,
		"Content-Type":"application/json"
	}

	body = {
		"documentId":doc,
		"compressionLevel":level	
	}

	request = requests.post(f"{HOST}/pdf-services/api/documents/modify/pdf-compress", json=body, headers=headers)
	return request.json()

def linearizePDF(doc, id, secret):

	headers = {
		"client_id":id,
		"client_secret":secret,
		"Content-Type":"application/json"
	}

	body = {
		"documentId":doc
	}

	request = requests.post(f"{HOST}/pdf-services/api/documents/optimize/pdf-linearize", json=body, headers=headers)
	return request.json()

def checkTask(task, id, secret):

	headers = {
		"client_id":id,
		"client_secret":secret,
		"Content-Type":"application/json"
	}

	done = False
	while done is False:

		request = requests.get(f"{HOST}/pdf-services/api/tasks/{task}", headers=headers)
		status = request.json()
		if status["status"] == "COMPLETED":
			done = True
			# really only need resultDocumentId, will address later
			return status
		elif status["status"] == "FAILED":
			print("Failure. Here is the last status:")
			print(status)
			sys.exit()
		else:
			print(f"Current status, {status['status']}, percentage: {status['progress']}")
			sleep(5)

def downloadResult(doc, path, id, secret):
	
	headers = {
		"client_id":id,
		"client_secret":secret
	}

	with open(path, "wb") as output:
		
		bits = requests.get(f"{HOST}/pdf-services/api/documents/{doc}/download", stream=True, headers=headers).content 
		output.write(bits)

input = "../../inputfiles/input.pdf"
print(f"File size of input: {os.path.getsize(input)}")
doc = uploadDoc(input, CLIENT_ID, CLIENT_SECRET)
print(f"Uploaded doc to Foxit, id is {doc['documentId']}")

task = compressPDF(doc["documentId"], "HIGH", CLIENT_ID, CLIENT_SECRET)
print(f"Created task, id is {task['taskId']}")

result = checkTask(task["taskId"], CLIENT_ID, CLIENT_SECRET)
print("Done converting to PDF. Now doing linearize.")

task = linearizePDF(result["resultDocumentId"], CLIENT_ID, CLIENT_SECRET)
print(f"Created task, id is {task['taskId']}")

result = checkTask(task["taskId"], CLIENT_ID, CLIENT_SECRET)
print("Done with linearize task.")

output = "../../output/really_optimized.pdf"
downloadResult(result["resultDocumentId"], output , CLIENT_ID, CLIENT_SECRET)
print(f"Done and saved to: {output}.")
print(f"File size of output: {os.path.getsize(output)}") 